In [152]:
#Import required libraries
import pandas as pd
import numpy as np

In [153]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

Load Dataset using Pandas Library

In [154]:
df = pd.read_csv("/content/diabetes.csv")
df.head(5)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


Analyze the dataset

In [155]:
print("Dataset Shape:")
print(df.shape)

Dataset Shape:
(768, 9)


In [156]:
print("Name of Columns:")
print(df.columns)

Name of Columns:
Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')


In [157]:
print("Data Types:")
print(df.dtypes)

Data Types:
Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


In [158]:
print("Missing Values:")
print(df.isnull().sum())

Missing Values:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


In [159]:
print("Summary of Dataset:")
df.describe()

Summary of Dataset:


,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


Separate Dataset into features and target variables

In [160]:
# Features
X = df.drop("Outcome", axis=1)

# Target
y = df["Outcome"]

print("Feature Shape:", X.shape)
print("Target Shape:", y.shape)

Feature Shape: (768, 8)
Target Shape: (768,)


In [161]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("Training dataset (features):", X_train.shape)
print("Training dataset (labels):", y_train.shape)
print("Testing dataset (features):", X_test.shape)
print("Testing dataset (labels):", y_test.shape)

Training dataset (features): (614, 8)
Training dataset (labels): (614,)
Testing dataset (features): (154, 8)
Testing dataset (labels): (154,)


Apply feature scaling using StandardScaler

In [162]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [163]:
print("First five rows of the standardized training dataset:")
print(X_train_scaled[:5])

print("\nFirst five rows of the standardized testing dataset:")
print(X_test_scaled[:5])

First five rows of the standardized training dataset:
[[-0.52639686 -1.15139792 -3.75268255 -1.32277365 -0.70120553 -4.13525578
  -0.49073479 -1.03594038]
 [ 1.58804586 -0.27664283  0.68034485  0.23350519 -0.70120553 -0.48916881
   2.41502991  1.48710085]
 [-0.82846011  0.56687102 -1.2658623  -0.09071957  0.01344832 -0.42452187
   0.54916055 -0.94893896]
 [-1.13052335  1.2541786  -1.04961706 -1.32277365 -0.70120553 -1.30372015
  -0.63929127  2.79212217]
 [ 0.68185612  0.41066475  0.57222224  1.07648956  2.48460077  1.83812075
  -0.68682934  1.13909516]]

First five rows of the standardized testing dataset:
[[ 0.68185612 -0.71402038 -0.61712658  0.81710976  0.93474906  0.26073561
  -0.11637247  0.87809089]
 [-0.52639686 -0.27664283  0.30191569  0.7522648  -0.70120553  0.48053518
  -0.954231   -1.03594038]
 [-0.52639686 -0.40160784 -0.29275872 -1.32277365 -0.70120553 -0.15300476
  -0.9245197  -1.03594038]
 [ 1.28598261 -0.43284909  0.57222224 -1.32277365 -0.70120553 -0.95462672
   1.1493

**PART 2** - Building ANN

In [164]:
import tensorflow as tf
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras import layers
from tensorflow.keras.layers import Dense, Conv2D, MaxPooling2D, Flatten, InputLayer

In [165]:
# Build a sequential neural network for diabetes prediction
model = Sequential([
    layers.Input(shape=(X_train_scaled.shape[1],), name="input_layer"),  # Accepts all input features from the processed dataset
    layers.Dense(units=64, activation='relu', name="hiddenlayer_1"),     # First hidden layer with 32 neurons using ReLU activation
    layers.Dense(units=32, activation='relu', name="hiddenlayer_2"),     # Second hidden layer with 16 neurons to learn deeper patterns
    layers.Dense(units=1, activation='sigmoid', name="output_layer")     # Produces the probability of the positive class
])

In [166]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [167]:
model.summary()

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ hiddenlayer_1 (Dense)           │ (None, 64)             │           576 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ hiddenlayer_2 (Dense)           │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output_layer (Dense)            │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,689 (10.50 KB)

 Trainable params: 2,689 (10.50 KB)

 Non-trainable params: 0 (0.00 B)

**Part 3 - Training the Model**

In [168]:
# Train the ANN model

history = model.fit(
    X_train_scaled,
    y_train,
    epochs=300,
    batch_size=32,
    validation_split=0.2,
    verbose=1
)

Epoch 1/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.6599 - loss: 0.6548 - val_accuracy: 0.6341 - val_loss: 0.6577
Epoch 2/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.6864 - loss: 0.5920 - val_accuracy: 0.6748 - val_loss: 0.6012
Epoch 3/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7189 - loss: 0.5427 - val_accuracy: 0.6911 - val_loss: 0.5577
Epoch 4/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7576 - loss: 0.5060 - val_accuracy: 0.7073 - val_loss: 0.5291
Epoch 5/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.7699 - loss: 0.4823 - val_accuracy: 0.7317 - val_loss: 0.5086
Epoch 6/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7923 - loss: 0.4651 - val_accuracy: 0.7480 - val_loss: 0.4897
Epoch 7/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.7821 - loss: 0.4539 - val_accuracy: 0.7398 - val_loss: 0.4851
Epoch 8/300
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7882 - loss: 0.4451 - val_accuracy: 0.

In [179]:
# Calculate the accuracy achieved at the end of training
final_accuracy = history.history['accuracy'][-1]

print("Training accuracy acheived by the Model:", round(final_accuracy * 100, 2), "%")

Training accuracy acheived by the Model: 100.0 %


In [170]:
print("Total epochs executed:", len(history.history['accuracy']))

Total epochs executed: 300


**Part 4 - Model Evaluation**

In [171]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [172]:
# Obtain probability scores for each test instance
y_pred_prob = model.predict(X_test_scaled)

# Display the predicted probabilities of the first five instances
print(y_pred_prob[:5])

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step
[[0.4713235 ]
 [0.04025404]
 [0.0008943 ]
 [0.17844182]
 [0.0509356 ]]


In [173]:
# Classify each prediction based on a 0.5 decision threshold
y_pred = (y_pred_prob >= 0.5).astype(int)

In [174]:
accuracy = accuracy_score(y_test, y_pred)

print("Model Accuracy:", round(accuracy * 100, 2), "%")

Model Accuracy: 74.03 %


In [175]:
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[80 19]
 [21 34]]


In [180]:
import pandas as pd

cm_df = pd.DataFrame(
    cm,
    index=["True Class:No Diabetes", "True Class:Diabetes"],
    columns=["Model Prediction:No Diabetes", "Model Prediction:Diabetes"]
)

cm_df

,Model Prediction:No Diabetes,Model Prediction:Diabetes
True Class:No Diabetes,80,19
True Class:Diabetes,21,34


## Confusion Matrix Summary

The confusion matrix indicates that the model performed well in distinguishing between diabetic and non-diabetic patients.
* Out of 99 patients who did not have diabetes, the model correctly classified 80 as non-diabetic, while 19 were incorrectly predicted as diabetic (false positives).
* Similarly, among 55 patients with diabetes, the model correctly identified 34 cases, whereas 21 diabetic patients were incorrectly classified as non-diabetic (false negatives).

Overall, the model made 114 correct predictions out of 154 test samples, corresponding to an accuracy of approximately 74%. The results show that the model is more effective at identifying patients without diabetes than detecting those with diabetes. Reducing the number of false negatives would further improve the model, as missed diabetes cases can have important clinical implications.